packages/teuchos/parameterlist/src/Teuchos_ParameterListModifier.hpp

```diff
@@ -0,0 +1,243 @@
+//
+//                    Teuchos: Common Tools Package
+//                 Copyright (2004) Sandia Corporation
+//
+// Under terms of Contract DE-AC04-94AL85000, there is a non-exclusive
+// license for use of this work by or on behalf of the U.S. Government.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Michael A. Heroux (maherou@sandia.gov)
+//
+// ***********************************************************************
+// @HEADER
+
+#ifndef TEUCHOS_ParameterListModifier_hpp_
+#define TEUCHOS_ParameterListModifier_hpp_
+
+/*! \file Teuchos_ParameterListModifier.hpp
+    \brief Parameter List Modifier class
+*/
+
+#include "Teuchos_Describable.hpp"
+#include "Teuchos_ParameterEntryValidator.hpp"
+#include "Teuchos_map.hpp"
+#include "Teuchos_RCP.hpp"
+namespace Teuchos {
+
+
+#ifndef DOXYGEN_SHOULD_SKIP_THIS
+class ParameterList;
+#endif
+
+/** \brief Abstract interface for an object that can modify both a
+ *  parameter list and the parameter list being used during the
+ *  validation stage.
+ *
+ * A parameter (sub)list modifier can be used to process optional fields and
+ * dependent fields before validation.  It can also be used after validation
+ * to reconcile parameters that may have dependencies on other parameters.
+ */
+class TEUCHOSPARAMETERLIST_LIB_DLL_EXPORT ParameterListModifier : public Describable {
+
+public:
+
+  //! @name Constructors/Destructor
+  //@{
+
+  //! Constructor
+  ParameterListModifier();
```

Comment:
We should set up the class so we can write `=default` and not have to implement the default constructor elsewhere.  See below.